src/TrajectoryPlanner/src/TrajectoryGenerator.cpp

```diff
@@ -853,3 +853,104 @@ bool TrajectoryGenerator::getDesiredZMPPosition(std::vector<iDynTree::Vector2> &
+    double leftYawDeltaInRad, rightYawDeltaInRad;
+    // left foot
+    measuredPositionLeft(0) = m_measuredTransformLeft.getPosition()(0);
+    measuredPositionLeft(1) = m_measuredTransformLeft.getPosition()(1);
+    measuredAngleLeft = m_measuredTransformLeft.getRotation().asRPY()(2);
+    leftYawDeltaInRad = m_leftYawDeltaInRad;
+    // right foot
+    measuredPositionRight(0) = m_measuredTransformRight.getPosition()(0);
+    measuredPositionRight(1) = m_measuredTransformRight.getPosition()(1);
+    measuredAngleRight = m_measuredTransformRight.getRotation().asRPY()(2);
+    rightYawDeltaInRad = m_rightYawDeltaInRad;
+    correctLeft = m_correctLeft;
+    m_newFreeSpaceEllipse = false;
+    nominalWidth = m_nominalWidth;
+
+    iDynTree::Vector2 measuredPosition;
+    double measuredAngle;
+    measuredPosition = correctLeft ? measuredPositionLeft : measuredPositionRight;
+    measuredAngle = correctLeft ? measuredAngleLeft : measuredAngleRight;
+
+    Eigen::Vector2d unicyclePositionFromStanceFoot, footPosition, unicyclePosition, unicycleOdom;
+    unicyclePositionFromStanceFoot(0) = 0.0;
+    Eigen::Matrix2d unicycleRotation; //rotation world -> unicycle
+    double unicycleAngle, unicycleAngleOdom;
+
+    if (stanceFoot == "left")
+    {
+        unicyclePositionFromStanceFoot(1) = -nominalWidth/2;
+        unicycleAngle = measuredAngleLeft;  //- leftYawDeltaInRad
+        unicycleAngleOdom = measuredAngleLeft - leftYawDeltaInRad;
+        footPosition = iDynTree::toEigen(measuredPositionLeft);
+        //stanceFoot = "left";
+    }
+    else if (stanceFoot == "right")
+    {
+        unicyclePositionFromStanceFoot(1) = nominalWidth/2;
+        unicycleAngle = measuredAngleRight;    //- rightYawDeltaInRad
+        unicycleAngleOdom = measuredAngleRight - rightYawDeltaInRad;
+        footPosition = iDynTree::toEigen(measuredPositionRight);
+        //stanceFoot = "right";
+    }
+    else
+    {
+        return false;
+    }
+
+    //Odom (expressed in odom frame)
+    double s_theta = std::sin(unicycleAngleOdom);
+    double c_theta = std::cos(unicycleAngleOdom);
+    unicycleRotation(0,0) = c_theta;
+    unicycleRotation(0,1) = -s_theta;
+    unicycleRotation(1,0) = s_theta;
+    unicycleRotation(1,1) = c_theta;
+    unicyclePosition = unicycleRotation * unicyclePositionFromStanceFoot;   // position in respect to the support foot
+    unicycleOdom = unicyclePosition + footPosition;     //postion in odom frame
+    virtualUnicyclePose(0) = unicycleOdom(0);
+    virtualUnicyclePose(1) = unicycleOdom(1);
+    virtualUnicyclePose(2) = unicycleAngleOdom;
+
+    iDynTree::Vector2 referencePointInAbsoluteFrame;
+    iDynTree::toEigen(referencePointInAbsoluteFrame) = unicycleRotation * (iDynTree::toEigen(m_referencePointDistance)) + unicycleOdom;
+    referenceUnicyclePose(0) = referencePointInAbsoluteFrame(0);
+    referenceUnicyclePose(1) = referencePointInAbsoluteFrame(1);
+    referenceUnicyclePose(2) = unicycleAngleOdom;
+    
```

Comment:
Maybe this can be handy: https://github.com/robotology/unicycle-footstep-planner/blob/f9db393b73ef92b2bf93ceac91da7a1db607f707/include/UnicycleFoot.h#L60